### PR TITLE
Add unit tests for DepQuant functions and refactor findFirstPos

### PR DESCRIPTION
**Pull request by athulya-arm**
- Add unit test for DepQuant functions `State::updateStates`, `State::updateStatesEOS`, `State::checkAllRdCostsOdd1`, `findFirstPos`.
- Modify the DepQuant constructor to include a new `enableOpt` parameter to create a reference object and an optimized object for testing.
- Move `m_updateStates`, `m_updateStatesEOS`, `m_checkAllRdCostsOdd1`, `m_findFirstPos` from the private to the public section of the class to enable unit test access.
- Add /wd4100 to unit test MSVC flags to suppress unreferenced parameter warnings from Quant.h, matching the existing library build configuration.
- Extract the scalar findFirstPos logic from `xQuantDQ` into a dedicated function `findFirstPos`.
- Call `m_findFirstPos` unconditionally from `xQuantDQ` and move the final coefficient-wise refinement from `xQuantDQ` into `findFirstPos` for both the C and x86 implementations.

### CODE REVIEW

**Comment by adamjw24** [comment]
If you are adding a reference C implementation of `findFirstPos`, maybe we should just use that? What I mean is that the virtual implementation is supposed to find SBB-exact pointer to the first codec coefficient and than `xQuantDq` refines that to coefficient exact resolution. If we have a C reference implementation of `findFirstPos`, maybe it would be just easier to say the function returns a cofficient exact first pos. This would require to put the refinement at the end of the SIMD as well as C implementations, and to remove it from `xQuantDq`. The nullptr-check if `xQuantDq` could also be removed.

File: source/Lib/CommonLib/DepQuant.cpp [hunk 1]
```diff
@@ -55,6 +55,69 @@ namespace vvenc {
...
+          break;
+        }
+      }
+      if( firstTestPos >= 0 )
+      {
+        // If a coefficient was found, advance the pointer to the end of the current subblock
+        // for the subsequent coefficient-wise refinement.
+        firstTestPos += sbbSize - 1;
+      }
+    }
```
Comment (adamjw24) [inline source/Lib/CommonLib/DepQuant.cpp]: I think all of this is obsolete, no?

Comment (athulya-arm) [reply source/Lib/CommonLib/DepQuant.cpp]: Do you mean the whole function or the newly added logic inside it?
We meant adding a scalar reference C would be useful to serve as a generic fallback, a readable reference for future SIMD implementations like Neon, rather than relying on the x86 version to understand the logic and also as a baseline for unit tests to compare optimized implementations against.

Comment (adamjw24) [reply source/Lib/CommonLib/DepQuant.cpp]: The newly added logic - I did ask for the function previously myself.

I guess you are right, its not necessary but doesn't hurt either. Could you maybe add a comment that its just a reference for SIMD and that the following loop would also take care of that?

Comment (athulya-arm) [reply source/Lib/CommonLib/DepQuant.cpp]: Thanks, I removed the scalar 4x4 prescan from `findFirstPos` to avoid confusion. Also updated the unit test for "not-found cases" to check for a negative pos, which is all `xQuantDQ` depends on.

File: test/vvenc_unit_test/vvenc_unit_test.cpp [hunk 2]
```diff
@@ -253,6 +254,417 @@ class DimensionGenerator
...
+        state.ctx.cff[j] = rng.get( state.cffBitsCtxOffset, state.cffBitsCtxOffset + 4 );
+
+        for( int k = 0; k < DQIntern::RateEstimator::sm_maxNumSigCtx; ++k )
+        {
+          sigBuf[j][k].intBits[0] = g20();
+          sigBuf[j][k].intBits[1] = g20();
+        }
+        decisions_ref.rdCost[j] = decisions_opt.rdCost[j] = DQIntern::rdCostInit >> 2;
+        decisions_ref.absLevel[j] = decisions_opt.absLevel[j] = -1;
+        decisions_ref.prevId[j] = decisions_opt.prevId[j] = -2;
```
Comment (adamjw24) [inline test/vvenc_unit_test/vvenc_unit_test.cpp]: `m_checkAllRdCostsOdd1` if the trellis has already been started, i.e. prevId != -2.

The only function that can correctly start a trellis path is `checkRdCostStart`. Sorry about that, this one you really just have to know....

The easy solution would be to just initialize prevId (at least for some states) to [0...3] and absLevel to a positive integer.

Also, an interesting test would be to initlize state 0 and 2 to prevId [0...3] prevId and positive absLevel and have states 1 and 3 initialized as you did - this way correct handling of mixed valid and invalid paths can be tested.

File: test/vvenc_unit_test/vvenc_unit_test.cpp [hunk 3]
```diff
@@ -253,6 +254,417 @@ class DimensionGenerator
...
+    scanInfo.insidePos = rng.get( 0, 15 );
+    scanInfo.nextInsidePos = rng.get( 0, 15 );
+    scanInfo.gtxCtxOffsetNext = ( int8_t )rng.get( 0, 200 );
+    scanInfo.sigCtxOffsetNext = ( int8_t )rng.get( 0, 200 );
+    scanInfo.currNbInfoSbb.numInv = ( uint8_t )rng.get( 0, 5 );
+    std::generate( scanInfo.currNbInfoSbb.invInPos, scanInfo.currNbInfoSbb.invInPos + 5, g4 );
+
+    for( int idx = 0; idx < 4; ++idx )
+    {
+      decisions.prevId[idx] = rng.getOneOf<int8_t>( { -1, 0, 1, 2, 3 } );
```
Comment (adamjw24) [inline test/vvenc_unit_test/vvenc_unit_test.cpp]: Keep testing -2, so that we also keep track of invalid (or - not yet started), trellis paths.

File: test/vvenc_unit_test/vvenc_unit_test.cpp [hunk 4]
```diff
@@ -253,6 +254,417 @@ class DimensionGenerator
...
+
+  bool passed = true;
+
+  for( unsigned i = 0; i < num_cases; ++i )
+  {
+    std::memset( &curr_ref, 0xCD, sizeof( curr_ref ) );
+    std::memset( &curr_opt, 0xCD, sizeof( curr_opt ) );
+
+    scanInfo.scanIdx = rng.get( 16, 128, 16 );
+    scanInfo.sbbSize = rng.getOneOf<int8_t>( { 4, 8, 16 } );
```
Comment (adamjw24) [inline test/vvenc_unit_test/vvenc_unit_test.cpp]: Does sbbSize of 8 (i.e. 4x2 or 2x4) exist? I though its either 2x2 or 4x4 for coefficient subblocks.

**Comment by adamjw24** [comment]
Thank you for including the tests. This is very low level code so its full of preassumptions that are only clear if you have in-depth knowledge of the VVC standard.

I have some general comments here. Using random number generator to generate the rdCosts for the `checkRdCost*` functions might make the values so far apart, that it would not show if decisions were based on faulty calcualtions (just because the values are not close-by, and the output is always only the best found cost - the dismissed options are not cross-checked).

Further, there are really many cases that can, and eventually will, arise. Here, I would propose a different testing scheme, more like integration tests. I.e. generate different "realistic" residual signals (e.g. by generating random or even non-random signals, than filtering them somehow, and using the difference as residual), applying the forward transform, quantization, and than checking the resulting transform coefficients. Do that at a wide range of QPs (e.g. 0-50 in steps of x) for a wide range of blocks sizes and transforms (to test zero-out handling) and most cases should be tested. The biggest issue here would be generating apropriate residuals.

If the current way of testing (unit) works better for you, I guess thats also ok, but I see many cases that are not caught with the approach, as commented above. Once you have some implementation I would also test some corpus of sequences at wide range of QPs versus the scalar implementation and see if the bitstreams are bit-equal. This is ultimately what we did when developing the x86 simd. Anyway, I am ok to proceed as is, but would be great if at least the issues mentioned were resolved.

**Comment by athulya-arm** [comment]
Thank you for the feedback. For this patch, we plan to strengthen the existing unit test coverage rather than adding integration style transform+quant tests. The current unit tests were already useful in bringing up the Neon implementation and we will extend them with cases like
1. tie/near-tie cases with close state.rdCost values and small pq_a_dist / pq_b_dist deltas, so a small arithmetic difference can change the selected path.
2. SCAN_EOCSBB cases with state.numSig == 0
We will also address the other review comments.

Also, following the suggestion to add testing prevId == -2 in check_updateStates, I noticed that the C and x86 paths behave differently for that case. The C path bypasses the entire update for that state, while the x86 path still updates/clears some fields. Since this case fails the comparison, I think it would be good to clarify whether this is the intended handling for an invalid path.

**Comment by adamjw24** [comment]
> Also, following the suggestion to add testing prevId == -2 in check_updateStates, I noticed that the C and x86 paths behave differently for that case. The C path bypasses the entire update for that state, while the x86 path still updates/clears some fields. Since this case fails the comparison, I think it would be good to clarify whether this is the intended handling for an invalid path.

Thats what I mean when mentioning integration tests. The update in x86 for prevId=-2 are irrelevant, as long as they are correctly overwritten later. The values are only updated because of how SIMD works, and if the values written back are not reset to the inital state, its only because its irrelevant for those values.

Please give me more info on the problematic values than we can discuss the best way to resolve that.

**Comment by athulya-arm** [comment]
C: https://github.com/fraunhoferhhi/vvenc/blob/master/source/Lib/CommonLib/DepQuant.cpp#L893
X86: https://github.com/fraunhoferhhi/vvenc/blob/master/source/Lib/CommonLib/x86/DepQuantX86.h#L72

For prevId == -2, the x86 path processes 4 lanes (stateId) together and still writes the corresponding lane for the invalid state, looks like a reset, while the C path bypasses the update for that state.
Pseudo-code of the resulting x86 state for that lane:
```
curr.numSig[stateId] = 0;
curr.refSbbCtxId[stateId] = -1;
curr.remRegBins[stateId] = curr.initRemRegBins;
curr.tplAcc[0..15][stateId] = 0;
curr.sum1st[0..15][stateId] = 0;
curr.ctx.cff[stateId] = scanInfo.gtxCtxOffsetNext;
curr.ctx.sig[stateId] = scanInfo.sigCtxOffsetNext;
```
If those values are indeed don’t-care for prevId == -2 and are always overwritten before becoming relevant, I can adjust the unit test accordingly.

**Comment by adamjw24** [comment]
> C: https://github.com/fraunhoferhhi/vvenc/blob/master/source/Lib/CommonLib/DepQuant.cpp#L893 X86: https://github.com/fraunhoferhhi/vvenc/blob/master/source/Lib/CommonLib/x86/DepQuantX86.h#L72
> 
> For prevId == -2, the x86 path processes 4 lanes (stateId) together and still writes the corresponding lane for the invalid state, looks like a reset, while the C path bypasses the update for that state. Pseudo-code of the resulting x86 state for that lane:
> 
> ```
> curr.numSig[stateId] = 0;
> curr.refSbbCtxId[stateId] = -1;
> curr.remRegBins[stateId] = curr.initRemRegBins;
> curr.tplAcc[0..15][stateId] = 0;
> curr.sum1st[0..15][stateId] = 0;
> curr.ctx.cff[stateId] = scanInfo.gtxCtxOffsetNext;
> curr.ctx.sig[stateId] = scanInfo.sigCtxOffsetNext;
> ```
> 
> If those values are indeed don’t-care for prevId == -2 and are always overwritten before becoming relevant, I can adjust the unit test accordingly.

- `numSig` ... is overwritten with the same value (`0`)
- `refSbbCtxId` ... is overwritten with same value (`-1`)
- `sum1st` ... is overwritten with the same value (`0`)
- `remRegBins` ... could already be initalized to `initRemRegBins` in the init stage (I guess its really only not because the value is initalized after `DQInter::initStates` call). Anyway, the first valid coefficient (`prevId == -1`) will correctly overwrite it anyway
- `tplAcc` ... will be correctly fully overwritten the first time a valid coefficient is processed (`prevId == -1`)
- `ctx.cff` ... is not accessed by `checkRdCostStart`, the only function that can produce valid decision for a `prevId == -2` state
- `ctx.sig` .... the same as `ctx.cff`

In that sense, changes `remRegBins`, `tplAcc`, `ctx.cff` and `ctx.sig` for invalid paths (`prevId == -2`) are irrelevant and don't need to be compared for any invalid paths.

The most important aspect is that the `rdCost` of a `prevId == -2` decision always has to be larger than any of the valid decisions (`prevId > -2`), so that for the next coefficients its never chosen as previous `stateId`, the same as with the state for which `prevId == -2` was chosen. So:
- after `checkRdCost*` the `Decisions::rdCost[i]` for each `i` where `prevId == -2` has to be larger than for any `i` for which  for which `prevId > -2`
- after `updateState{,EOS}`, the `StateMem::rdCost[i]` for each `i` where `Decisions::prevId[i] == -2` has to be larger than any for `i` for which which `prevId > -2`

**Comment by adamjw24** [comment]
Also, another aspect that might be important. `Decisions::prevId[i] == -2` might only occur at the beginning of the search when the `StateMem` is initialized with default values. Once any non-default values has been written to any of the states `Decisions::prevId[i] == -2` cannot occur. Its because for the first coefficients only transitions to states 0 and 2 are possible (because the trellis startes in state 0), thus states 1 and 3 are still invalid. But for the next coefficient, as we have valid values in previous states 0 and 2, we have valid transitions to any of the states 0...3 available.

File: test/vvenc_unit_test/vvenc_unit_test.cpp (hunk 2 above)
Comment (athulya-arm) [reply test/vvenc_unit_test/vvenc_unit_test.cpp]: `absLevel` and `prevId` in `checkAllRdCostsOdd1()` are output decision values, the function does not use them as inputs. In this test, we updated their initialization to invalid values so the unit test fails by default if either the reference or optimized implementation does not overwrite them correctly. The goal here is to verify that both implementations update these decision outputs consistently, rather than to model a pre-started trellis state.

File: test/vvenc_unit_test/vvenc_unit_test.cpp (hunk 4 above)
Comment (athulya-arm) [reply test/vvenc_unit_test/vvenc_unit_test.cpp]: Done

File: test/vvenc_unit_test/vvenc_unit_test.cpp (hunk 3 above)
Comment (athulya-arm) [reply test/vvenc_unit_test/vvenc_unit_test.cpp]: Done

**Comment by adamjw24** [comment]
Thank you for addressing the comments, I think its good to go now.

File: test/vvenc_unit_test/vvenc_unit_test.cpp [hunk 5]
```diff
@@ -253,6 +254,518 @@ class DimensionGenerator
...
+
+    curr_ref.initRemRegBins = 4;
+    std::generate( curr_ref.numSig, curr_ref.numSig + 4, g6 );
+    std::generate( curr_ref.remRegBins, curr_ref.remRegBins + 4, g11 );
+    for( int idx = 0; idx < 16; ++idx )
+    {
+      std::generate( curr_ref.tplAcc[idx], curr_ref.tplAcc[idx] + 4, g8 );
+      std::generate( curr_ref.sum1st[idx], curr_ref.sum1st[idx] + 4, g8 );
+    }
+    std::fill_n( &curr_ref.absVal[0][0], 16 * 4, 0 ); // AbsVal is always set to 0 prior to updateStates.
```
Comment (adamjw24) [inline test/vvenc_unit_test/vvenc_unit_test.cpp:486]: You could set `absVal[i][j]` to nonZero for `i > insidePos`, to ensure that they are shuffled properly (and zeroed for `prevId == -1`) during updateState.

File: test/vvenc_unit_test/vvenc_unit_test.cpp [hunk 6]
```diff
@@ -253,6 +254,518 @@ class DimensionGenerator
...
+
+    scanInfo.scanIdx = rng.get( 16, 128, 16 );
+    scanInfo.sbbSize = rng.getOneOf<int8_t>( { 4, 16 } );
+    scanInfo.numSbb = 16;
+
+    scanInfo.sbbPos = rng.get( 0, scanInfo.numSbb - 1 );
+    scanInfo.nextSbbRight = 0;
+    scanInfo.nextSbbBelow = 0;
+
+    scanInfo.insidePos = rng.get( 0, 15 );
```
Comment (adamjw24) [inline test/vvenc_unit_test/vvenc_unit_test.cpp:591]: For `updateStateEOS`, `insidePos` should always be `0`

File: test/vvenc_unit_test/vvenc_unit_test.cpp [hunk 7]
```diff
@@ -253,6 +254,518 @@ class DimensionGenerator
...
+    scanInfo.nextInsidePos = rng.get( 0, 15 );
```
Comment (adamjw24) [inline test/vvenc_unit_test/vvenc_unit_test.cpp:592]: For `updateStateEOS`, `nextInsidePos` should always be `16`

File: test/vvenc_unit_test/vvenc_unit_test.cpp [hunk 8]
```diff
@@ -253,6 +254,518 @@ class DimensionGenerator
...
+    scanInfo.gtxCtxOffsetNext = ( int8_t )rng.get( 0, 200 );
+    scanInfo.sigCtxOffsetNext = ( int8_t )rng.get( 0, 200 );
+    scanInfo.currNbInfoSbb.numInv = ( uint8_t )rng.get( 0, 5 );
```
Comment (adamjw24) [inline test/vvenc_unit_test/vvenc_unit_test.cpp:595]: For `insidePos == 0`, `scanInfo.currNbInfoSbb.numInv` is always equal to `0`

File: test/vvenc_unit_test/vvenc_unit_test.cpp [hunk 9]
```diff
@@ -253,6 +254,518 @@ class DimensionGenerator
...
+    std::generate( scanInfo.currNbInfoSbb.invInPos, scanInfo.currNbInfoSbb.invInPos + 5, g4 );
+
+    for( int idx = 0; idx < 4; ++idx )
+    {
+      // prevId >= 4 implies prevId - 4 == idx.
+      decisions.prevId[idx] = rng.getOneOf<int8_t>( { -1, 0, 1, 2, 3, int8_t( idx + 4 ) } );
```
Comment (adamjw24) [inline test/vvenc_unit_test/vvenc_unit_test.cpp:601]: `updateStateEOS` can also have a `prevId == -2`

**Comment by adamjw24** [comment]
> Thank you for addressing the comments, I think its good to go now.

Sorry, I did look into it a bit more and found minor comments. I guess its nothing critical, so let me know if you want to adapt it or should we merge it for now.

**Review by adamjw24 — COMMENTED** [review]
.

File: test/vvenc_unit_test/vvenc_unit_test.cpp [hunk 10]
```diff
@@ -253,6 +254,518 @@ class DimensionGenerator
...
+  {
+    std::memset( &curr_ref, 0xCD, sizeof( curr_ref ) );
+    std::memset( &curr_opt, 0xCD, sizeof( curr_opt ) );
+
+    scanInfo.insidePos = rng.get( 0, 15 );
+    scanInfo.nextInsidePos = rng.get( 0, 15 );
+    scanInfo.gtxCtxOffsetNext = ( int8_t )rng.get( 0, 200 );
+    scanInfo.sigCtxOffsetNext = ( int8_t )rng.get( 0, 200 );
+    scanInfo.currNbInfoSbb.numInv = ( uint8_t )rng.get( 0, 5 );
+    std::generate( scanInfo.currNbInfoSbb.invInPos, scanInfo.currNbInfoSbb.invInPos + 5, g4 );
```
Comment (adamjw24) [inline test/vvenc_unit_test/vvenc_unit_test.cpp:464]: Actually, just one last comment, to ensure we don't get undefined behavior. There are some constraints on the variables:
- `insidePos = ( nextInsidePos + 1 ) % sbbSize `
- all `scanInfo.currNbInfoSbb.invInPos` are smaller than `insidePos`
- `scanInfo.currNbInfoSbb.numInv <= std::min(insidePos, 5)`

If the tests function though, I guess its not super important to address this. But maybe the info can also be beneficial for you to develop the NEON SIMD (applying those restrictions will loosen the tests a bit because it would e.g. allow to write values in different order).

Comment (athulya-arm) [reply test/vvenc_unit_test/vvenc_unit_test.cpp]: Addressed all comments, please have a look.
Thank you.

File: test/vvenc_unit_test/vvenc_unit_test.cpp [hunk 11]
```diff
@@ -253,6 +254,540 @@ class DimensionGenerator
...
+    scanInfo.sbbSize = rng.getOneOf<int8_t>( { 4, 16 } );
+    scanInfo.nextInsidePos = rng.get( 0, scanInfo.sbbSize - 1 );
+    scanInfo.insidePos = ( scanInfo.nextInsidePos + 1 ) % scanInfo.sbbSize;
+    scanInfo.gtxCtxOffsetNext = ( int8_t )rng.get( 0, 200 );
+    scanInfo.sigCtxOffsetNext = ( int8_t )rng.get( 0, 200 );
+    scanInfo.currNbInfoSbb.numInv = ( uint8_t )rng.get( 0, std::min<int>( scanInfo.insidePos, 5 ) );
+
+    for( int inv = 0; inv < 5; ++inv )
+    {
+      scanInfo.currNbInfoSbb.invInPos[inv] = ( uint8_t )rng.get( 0, std::max<int>( 0, scanInfo.insidePos - 1 ) );
```
Comment (adamjw24) [inline test/vvenc_unit_test/vvenc_unit_test.cpp:472]: Very strictly, all `invInPos` (inverse template positions within the current subblock) are unique, while the code could generate the same position to be in the list more than once.

I think its a very minor issue, and I will take the code in for now - just keep that in mind for when you're developing the implementation.

**Comment by adamjw24** [comment]
Thank you for bearing with me through this long review process. I think the last minor issue is not super imporant, and I suspect anyway there is more small issues there that I just didn't catch. Please dont hesistate to reach out if you suspect that a problem during the NEON development is caused by the tests themselves and we can discuss here.

Also, please have a look at the following paper, where I documented some of the DQ implementation concepts used here that differ from the reference VTM implementation (like the inverse template).

A. Wieckowski, C. Stoffers, H. Schwarz, B. Bross and D. Marpe, "Fast Dependent Quantization Using Trellis Pruning, Forward Context Adaptation and Vectorization," 2024 32nd European Signal Processing Conference (EUSIPCO), Lyon, France, 2024, pp. 516-520, doi: 10.23919/EUSIPCO63174.2024.10714991.
